contrib/scalar/scalar.c

```diff
@@ -19,55 +19,98 @@ static int is_unattended(void) {
 	return git_env_bool("Scalar_UNATTENDED", 0);
 }
 
+/**
+ * Remove the deepest subdirectory in the provided path string. Path must not
+ * include a trailing path separator.
+ */
+static void strbuf_parentdir(struct strbuf *buf)
+{
+	size_t offset = offset_1st_component(buf->buf);
+	char *path_sep = find_last_dir_sep(buf->buf + offset);
+	strbuf_setlen(buf, path_sep ? path_sep - buf->buf : offset);
+}
+
+/**
+ * Given an absolute path at or below the enlistment root, find the
+ * enlistment root and working directory
+ */
+static int find_enlistment(struct strbuf *path,
+			   struct strbuf *enlistment_root)
+{
+	char *base_dir;
+	size_t offset = offset_1st_component(path->buf);
+	do {
+		const size_t len = path->len;
+
+		/* check if currently in enlistment root with src/ workdir */
+		strbuf_addstr(path, "/src/.git");
+		if (is_git_directory(path->buf)) {
+			strbuf_strip_suffix(path, "/.git");
+
+			if (enlistment_root) {
+				strbuf_addbuf(enlistment_root, path);
+				strbuf_parentdir(enlistment_root);
```

Comment:
Would it make sense to use `strbuf_add(enlistment_root, path, len);` here instead of copying the full `src` path and then finding its parent directory?

Reply:
Yes, that definitely simplifies the implementation. Updated!